File: Class/PhotoView/XHPhotoGroup.m
//
//  XHPhotoGroup.m
//  GMBuy
//
//  Created by chengxianghe on 15/9/21.
//  Copyright © 2015年 cn. All rights reserved.
//


#import "XHPhotoGroup.h"
#import "UIImageView+WebCache.h"
#import "XHPhotoBrowser.h"
#import "PhotoViewFrameHelper.h"

@interface XHPhotoGroup() <XHPhotoBrowserDelegate>
@property (nonatomic, strong) NSArray *photoItemArray;

@end

@implementation XHPhotoGroup
- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        // 清除图片缓存，便于测试
        //        [[SDWebImageManager sharedManager].imageCache clearDisk];
        
        //        NSLog(@"++++++++++++创建了9张图");
        for (int i = 0; i < 9; i ++) {
            UIImageView *imageV = [[UIImageView alloc] init];
            
            //            UIButton *btn = [[UIButton alloc] init];
            //让图片不变形，以适应按钮宽高，按钮中图片部分内容可能看不到
            imageV.contentMode = UIViewContentModeScaleAspectFill;
            imageV.clipsToBounds = YES;
            imageV.userInteractionEnabled = YES;
            UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(buttonClick:)];
            [imageV addGestureRecognizer:tap];
            imageV.tag = i;
            
            [self addSubview:imageV];
            
        }
        
    }
    return self;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    for (int i = 0; i < 9; i ++) {
        UIImageView *imageV = [[UIImageView alloc] init];
        
        //            UIButton *btn = [[UIButton alloc] init];
        //让图片不变形，以适应按钮宽高，按钮中图片部分内容可能看不到
        imageV.contentMode = UIViewContentModeScaleAspectFill;
        imageV.clipsToBounds = YES;
        imageV.userInteractionEnabled = YES;
        
        UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(buttonClick:)];
        [imageV addGestureRecognizer:tap];
        
        imageV.tag = i;
        
        [self addSubview:imageV];
    }
    
}

- (void)setXHPhotoGroupIsShowImage:(BOOL)isShowImage {
    
    // 刷新图片数据 (最多展示9张)
    
    /**
     *SDWebImageRetryFailed = 1<< 0,            默认选项，失败后重试
     *SDWebImageLowPriority = 1<< 1,            使用低优先级(列表用)
     *SDWebImageCacheMemoryOnly = 1<< 2,        仅仅使用内存缓存
     *SDWebImageProgressiveDownload = 1<< 3,    显示现在进度
     *SDWebImageRefreshCached = 1<< 4,          刷新缓存
     *SDWebImageContinueInBackground =1 << 5,   后台继续下载图像
     *SDWebImageHandleCookies = 1<< 6,          处理Cookie
     *SDWebImageAllowInvalidSSLCertificates= 1 << 7,    允许无效的SSL验证
     *SDWebImageHighPriority = 1<< 8,           高优先级(优先下载 移到队列最前)
     *SDWebImageDelayPlaceholder = 1<< 9        延迟显示占位图片
     */
    
    [self.subviews enumerateObjectsUsingBlock:^(UIImageView *imageView, NSUInteger idx, BOOL *stop) {
        
        if (idx < _photoItemArray.count) {
            
            if ([_photoItemArray[idx] isKindOfClass:[XHPhotoItem class]]) {
                
                if (isShowImage) {
                    
                    [self.subviews[idx].layer removeAnimationForKey:@"contents"];
                    
                    
                    [imageView sd_setImageWithURL:[NSURL URLWithString:[_photoItemArray[idx] thumbnail_pic]] placeholderImage:[UIImage imageNamed:@"whiteplaceholder"] options:SDWebImageAvoidAutoSetImage completed:^(UIImage *image, NSError *error, SDImageCacheType cacheType, NSURL *imageURL) {
                        
                        if (error != nil) {
                            [imageView setImage:[UIImage imageNamed:@"whiteplaceholder"]];
                            return ;
                        }
                        
                        if (!self.type.length && self.photoItemArray.count == 1) {
                            //                                    CGSize photoViewSize = [PhotoViewFrameHelper size:image.size MatchToSize:self.frame.size];
                            // super.frame = frame;
                            //                                    self.subviews[0].frame = (CGRect){0,0,photoViewSize};
//                            CGFloat width = image.size.width;
//                            CGFloat height = image.size.height;
//                            CGFloat scale = (height / width) / (imageView.bounds.size.height / imageView.bounds.size.width);
//                            if (scale < 0.99 || isnan(scale)) { // 宽图把左右两边裁掉
//                                imageView.contentMode = UIViewContentModeScaleAspectFill;
//                                imageView.layer.contentsRect = CGRectMake(0, 0, 1, 1);
//                            } else { // 高图只保留顶部
//                                imageView.contentMode = UIViewContentModeScaleToFill;
//                                imageView.layer.contentsRect = CGRectMake(0, 0, 1, width / height);
//                            }
                        }
                        
                        imageView.image = image;
                        
                        if (cacheType != SDImageCacheTypeMemory) {
                            CATransition *transition = [CATransition animation];
                            transition.duration = 0.15;
                            transition.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
                            transition.type = kCATransitionFade;
                            [imageView.layer addAnimation:transition forKey:@"contents"];
                        }
                    }];
                    
                } else {
                    [imageView setImage:[UIImage imageNamed:@"whiteplaceholder"]];
                }
            }
            
        }else {
            [imageView setImage:nil];
        }
        
    }];
    
    
    
}

- (void)setPhotoItemArray:(NSArray *)photoItemArray isShowImage:(BOOL)isShowImage
{
    if(![photoItemArray isKindOfClass:[NSArray class]]) {
        return;
    }
    _photoItemArray = photoItemArray;
    
    // 刷新图片布局
    [self setFrameSubviews];
    // 刷新图片数据
    [self setXHPhotoGroupIsShowImage:isShowImage];
}

- (void)setFrameSubviews {
    NSArray *array = [PhotoViewFrameHelper getPhotoViewFramesWithPhotoCount:self.photoItemArray.count andPhotoViewSize:self.frame.size];
    
    for (int i = 0; i < self.subviews.count; i ++) {
        UIView *imageV = self.subviews[i];
        
        if (i<self.photoItemArray.count) {
            imageV.frame = CGRectFromString(array[i]);
        } else {
            imageV.frame = CGRectZero;
        }
    }
}

- (void)buttonClick:(UITapGestureRecognizer *)sender
{
    //启动图片浏览器
    UIViewController *vc = (UIViewController *)[[self getParsentView:self] nextResponder];
    
    [XHPhotoBrowser showInView:self
                     container:vc
                  currentIndex:sender.view.tag
                    imageCount:self.photoItemArray.count
                      delegate:self];
    
//    XHPhotoBrowser *browserVc = [[XHPhotoBrowser alloc] init];
//    browserVc.sourceImagesContainerView = self; // 原图的父控件
//    browserVc.imageCount = self.photoItemArray.count; // 图片总数
//    browserVc.currentImageIndex = (int)sender.view.tag;
//    browserVc.delegate = self;
//    browserVc.containerVC = (UIViewController *)[[self getParsentView:self] nextResponder];
//    
//    UIImage *snapshotImage = [self.class snapshotImageAfterScreenUpdates:YES inView:[UIApplication sharedApplication].keyWindow];
//    browserVc.snapImage = snapshotImage;
//
//    [self.window.rootViewController presentViewController:browserVc animated:NO completion:nil];

}
- (UIView *)getParsentView:(UIView *)view{
    if ([[view nextResponder] isKindOfClass:[UIViewController class]] || view == nil) {
        return view;
    }
    return [self getParsentView:view.superview];
}

#pragma mark - photobrowser代理方法
// 返回临时占位图片（即原来的小图）
- (UIImage *)photoBrowser:(XHPhotoBrowser *)browser placeholderImageForIndex:(NSInteger)index
{
    if ([self.subviews[index] isKindOfClass:[UIImageView class]]) {
        return [(UIImageView*)self.subviews[index] image];
    } else {
        return [self.subviews[index] currentImage];
    }
    
}

// 返回高质量图片的url
- (NSURL *)photoBrowser:(XHPhotoBrowser *)browser highQualityImageURLForIndex:(NSInteger)index
{
    NSString *urlStr = [self.photoItemArray[index] original_pic];
    return [NSURL URLWithString:urlStr];
}

- (void)photoBrowser:(XHPhotoBrowser *)browser willShowAtIndex:(NSInteger)index {
    [(UIView *)self.subviews[index] setHidden:YES];
    NSLog(@"%s", __func__);
}

- (void)photoBrowser:(XHPhotoBrowser *)browser didShowAtIndex:(NSInteger)index {
    [(UIView *)self.subviews[index] setHidden:NO];
    NSLog(@"%s", __func__);

}

- (void)photoBrowser:(XHPhotoBrowser *)browser willDismissAtIndex:(NSInteger)index {
    [(UIView *)self.subviews[index] setHidden:YES];
    NSLog(@"%s", __func__);

}

- (void)photoBrowser:(XHPhotoBrowser *)browser didDismissAtIndex:(NSInteger)index {
    [(UIView *)self.subviews[index] setHidden:NO];
    NSLog(@"%s", __func__);

}

@end

File: XHPhotoBrowser/PhotoView/XHPhotoGroup.swift
//
//  XHPhotoGroup.swift
//  XHPhotoBrowser
//
//  Created by chengxianghe on 16/8/14.
//  Copyright © 2016年 cn. All rights reserved.
//

import UIKit

class XHPhotoGroupItem: NSObject {
    /// 缩略图
    var thumbnail_pic = ""
    
    /// 原图
    var original_pic = ""
    
    /// size
    var photoSize = CGSize.zero
}
typealias XHPhotoGroupShowClosure = (XHPhotoBrowser) -> (Void)

class XHPhotoGroup: UIView {

    fileprivate var imageViews = [UIImageView]()
    
    var photoBrowserDidShowClosure: XHPhotoGroupShowClosure?
    var photoBrowserDidDismissClosure: XHPhotoGroupShowClosure?
    
    var photoItemArray: [XHPhotoGroupItem]? {
        didSet {
            setupPhotoData()
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupPhotoGroup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setupPhotoGroup()
    }
    
    func setupPhotoGroup() {
        for _ in 0..<9 {
            let imageView = UIImageView()
            imageView.contentMode = .scaleAspectFill
            imageView.clipsToBounds = true
            imageView.isUserInteractionEnabled = true
            
            let tap = UITapGestureRecognizer.init(target: self, action: #selector(self.tapImage(sender:)))
            imageView.addGestureRecognizer(tap)
            self.addSubview(imageView)
            imageViews.append(imageView)
        }
    }
    
    func setupPhotoData() {        
        // 刷新图片布局
        self.setFrameSubviews()
        // 刷新图片数据
        self.setPhotoGroupImage()

    }
    
    func setFrameSubviews() {
        let array = PhotoViewFrameHelper.getPhotoViewFramesWithPhotoCount(count: photoItemArray?.count ?? 0, photoViewSize: self.frame.size, gap: 10)
        
        let enumer = imageViews.enumerated()
        for (index, imageView) in enumer {
            if index < array.count {
                imageView.frame = CGRectFromString(array[index])
            } else {
                imageView.frame = CGRect.zero;
            }
        }
    }
    
    func setPhotoGroupImage() {
        
        let enumer = imageViews.enumerated()
        for (index, imageView) in enumer {
            if index >= photoItemArray?.count ?? 0 {
                imageView.image = nil;
                return
            }
            
            imageView.layer.removeAnimation(forKey: "contents")
            
            let photoItem = photoItemArray![index]
            
            imageView.yy_setImage(with: NSURL(string: photoItem.original_pic)! as URL,
                                         placeholder: UIImage(named: "whiteplaceholder"),
                                         options: YYWebImageOptions.avoidSetImage,
                                         progress: nil,
                                         transform: nil,
                                         completion: {[weak imageView] (image, url, from, stage, error) in

                                            guard let imageView = imageView else {
                                                return
                                            }
                                            
                                            if (error != nil || image == nil) {
                                                imageView.image = UIImage(named: "whiteplaceholder")
                                                return ;
                                            }
                                            
                                            imageView.image = image;
                                            
                                            let width = image!.size.width * UIScreen.main.scale
                                            let height = image!.size.height * UIScreen.main.scale
                                            //
                                            let scale = (height / width) / (imageView.bounds.size.height / imageView.bounds.size.width);
                                            if (scale < 0.99 || scale.isNaN) { // 宽图把左右两边裁掉
//                                                imageView.alignTop = false
                                                imageView.contentMode = UIViewContentMode.scaleAspectFill;
                                                imageView.layer.contentsRect = CGRect(x: 0, y: 0, width: 1, height: 1);
                                            } else { // 高图只保留顶部
//                                                imageView.alignTop = true
                                                imageView.contentMode = UIViewContentMode.scaleToFill;
                                                let imageViewScale = imageView.bounds.size.width / imageView.bounds.size.height
                                                imageView.layer.contentsRect = CGRect(x: 0, y: 0, width: 1, height: width / height / imageViewScale);
                                            }
                                            
                                            if (from != YYWebImageFromType.memoryCacheFast) {
                                                let transition = CATransition()
                                                transition.duration = 0.15;
                                                transition.timingFunction = CAMediaTimingFunction(name:kCAMediaTimingFunctionEaseOut);
                                                transition.type = kCATransitionFade;
                                                imageView.layer.add(transition, forKey: "contents")
                                            }
        
            })
            
        }

    }
    
    @objc func tapImage(sender: UITapGestureRecognizer) {
        //启动图片浏览器
        let vc = self.xh_viewController
        let currentPage = imageViews.index(of: sender.view as! UIImageView) ?? 0

        var items = [XHPhotoItem]()
        let enumer = photoItemArray!.enumerated()
        
        for (index, photoItem) in enumer {
            let item = XHPhotoItem()
            item.thumbView = imageViews[index]
            item.largeImageURL = URL(string: photoItem.original_pic)
            item.shouldClipToTop = self.shouldClippedToTop(view: item.thumbView)
            item.caption = caption[Int(arc4random())%10];
            items.append(item)
        }

        let v = XHPhotoBrowser.init(groupItems: items)
        v.delegate = self
        v.fromItemIndex = currentPage
        v.blurEffectBackground = false
        v.hideToolBar = false
        v.pager.hidesForSinglePage = true
        v.singleTapOption = .dismiss
//        v.showToolBarWhenScroll = false
//        v.showCaptionWhenScroll = false
        
        //iPhone X上可以设置额外的选项
        v.isFullScreenWord = false
        v.isFullScreen = true
        
        v.show(inContaioner: (vc?.tabBarController!.view)!, animated: true, completion: nil)

    }

    func shouldClippedToTop(view: UIView?) -> Bool {
        if (view != nil) {
            if (view!.layer.contentsRect.size.height < 1) {
                return true;
            }
        }
        return false;
    }

}

extension XHPhotoGroup: XHPhotoBrowserDelegate {
    
    func xh_photoBrowserWillMove(toSuperView photoBrowser: XHPhotoBrowser) {
        let index = photoBrowser.fromItemIndex
        self.imageViews[index].isHidden = true
    }
    
    func xh_photoBrowserWillRemove(fromSuperView photoBrowser: XHPhotoBrowser) {
        let index = photoBrowser.currentPage
        self.imageViews[index].isHidden = false
    }
    
    func xh_photoBrowserWillDisplay(_ photoBrowser: XHPhotoBrowser) {
        print("将要展示")
    }
    
    func xh_photoBrowserDidDisplay(_ photoBrowser: XHPhotoBrowser) {
        print("已经展示")
        photoBrowserDidShowClosure?(photoBrowser)
    }
    
    func xh_photoBrowserWillDismiss(_ photoBrowser: XHPhotoBrowser) {
        print("将要消失")
    }
    
    func xh_photoBrowserDidDismiss(_ photoBrowser: XHPhotoBrowser) {
        print("已经消失")
        photoBrowserDidDismissClosure?(photoBrowser)
    }
    
    func xh_photoBrowser(_ photoBrowser: XHPhotoBrowser, didDisplayingImageAt index: Int, from fromIndex: Int) {
        print("正在展示第 \(index) 页 , fromIndex:\(fromIndex)")
        if fromIndex != NSNotFound {
            self.imageViews[fromIndex].isHidden = false
        }
        self.imageViews[index].isHidden = true

    }
}


